### ISSUE TITLE
problems with compiling csound web

### ISSUE THREAD

**Issue by docb** [csound/csound#1476]
i get the following error when compiling csound web:
first there is a warning:

```
+ emcc -s LINKABLE=1 -s ASSERTIONS=0 ../src/FileList.c -Iinclude -o FileList.bc
emcc: warning: generating an executable with an object extension (.bc).  If you meant to build an object file please use `-c, `-r`, or `-shared` [-Wemcc]
+ emcc -s LINKABLE=1 -s ASSERTIONS=0 ../src/CsoundObj.c -I../../include -Iinclude -o CsoundObj.bc
emcc: warning: generating an executable with an object extension (.bc).  If you meant to build an object file please use `-c, `-r`, or `-shared` [-Wemcc]
+ emcc -v -O3 -flto -DINIT_STATIC_MODULES=0 -s WASM=1 -s ASSERTIONS=1 -s LINKABLE=1 -s RESERVED_FUNCTION_POINTERS=1 -s TOTAL_MEMORY=268435456 -s ALLOW_MEMORY_GROWTH=1 -s NO_EXIT_RUNTIME=1 -s SINGLE_FILE=1 --pre-js ../src/FileList.js -s BINARYEN_ASYNC_COMPILATION=1 -s MODULARIZE=1 -s 'EXPORT_NAME="libcsound"' -s 'EXTRA_EXPORTED_RUNTIME_METHODS=["FS", "ccall", "cwrap"]' -s ENVIRONMENT=web CsoundObj.bc FileList.bc libcsound.a ../deps/lib/libsndfile.a ../deps/lib/libogg.a ../deps/lib/libvorbis.a ../deps/lib/libvorbisenc.a ../deps/lib/libFLAC.a -o libcsound.js
```
then the error (seems to be related):
```
wasm-ld: error: unknown file type: CsoundObj.bc
wasm-ld: error: unknown file type: FileList.bc
emcc: error: [..]
```
(emscripten: 2.0.18, os ubuntu 18.04 csound: e6c9ce1d1f20cb2a688402be089961f3583d2b37)

**Comment by kunstmusik** [comment]
We have not been using Emscripten for current work and haven't tested that build in some time. I imagine the issue is related to newer versions of Emscripten as the docker-based Emscripten build is still building with Github Actions.  We've been focused mostly on the browser build in the feature/webaudio-csound branch which uses nix to build the latest. 

For now I'd suggest using an older version of Emscripten or trying out the build in feature/webaudio-csound.

**Comment by vlazzarini** [comment]
Interesting, how is webcsound to be built in the new versions? What's the toolchain? Is it not using CMake anymore?

**Comment by kunstmusik** [comment]
@hlolli could explain better, but most of the work is in nix packages that
define how to build each piece:

https://github.com/csound/csound/tree/feature/webaudio-csound/wasm/src

This isn't using cmake but it could do so.

**Comment by vlazzarini** [comment]
Thx. Perhaps @hlolli could list the toolchain somewhere in the README so we can try out the build. I see mentions to yarn but not sure what else is needed.

**Comment by kunstmusik** [comment]
Definitely needs improved instructions. For the moment, it's pretty much:

1. download yarn
2. download nix (https://nixos.org/download.html, the curl command works
fine on mac)
3. cd wasm
4. yarn build    (builds the csound-wasm module, which is just libcsound
compiled to wasm)
5. yarn link
6. cd browser
7. yarn link csound  (tells to use the local version of csound-wasm rather
than the one on npm)
8. yarn build     (build the csound-browser module)

**Comment by vlazzarini** [comment]
thanks, that should get me going. Is the compiler supplied by nixos then? It would bd interesting to compare performance vs emcc

**Comment by docb** [comment]
thanks with emsdk 1.40.1 it works.

**Comment by kunstmusik** [comment]
@vlazzarini Yes, nix builds all of the dependencies to make builds completely reproducible. The downside is that first time builds can be very slow, but after that it's fine as the dependencies don't change.

**Comment by docb** [comment]
i have have managed to compile with nix with the following hints/issues/questions:

- after step 6: `npm install ` seems to be nesscesary


- here is some code which worked well with CsoundObj.js (6.15) in a gatsby/react-bootstrap/webpack environment

```
  componentDidMount() {
      let csjs = document.createElementNS("http://www.w3.org/1999/xhtml", "script");
      csjs.type = "text/javascript";
      csjs.src = withPrefix("/CsoundObj.js");
      document.getElementsByTagName("head")[0].appendChild(csjs)
      let self = this;
      csjs.onload = function() {
         window.CsoundObj.initialize().then(() => {
           console.log("loaded WASM runtime");
           self.csound = new window.CsoundObj(); // all ok ready to go ....
           //........
```

now i tried to transfer to version 6.16

```
  componentDidMount() {
      let self = this;
      const url = withPrefix('csound.dev.esm.js');
      import(/* webpackIgnore: true */ url).then(module => {
        console.log(module); // seems to be OK
        module.Csound().then(CsoundObj => {  // fails
           self.csound = CsoundObj;
           //........
```

but here i get on chromium:

```
Module
  Csound: (...)
  default: (...)
  Symbol(Symbol.toStringTag): "Module"
  get Csound: ƒ ()
  set Csound: ƒ ()
  get default: ƒ ()
  set default: ƒ ()  index.js:67 

index.js Single Thread AudioWorklet
worklet.singlethread.main.js:41 WorkletSinglethread_Main Initialize Module
worklet.singlethread.main.js:175 WorkletSinglethread_Main wrapping Comlink proxy endpoint on the audioWorkletNode.port
../src/workers/worklet.singlethread.worker.js:73 WorkletSinglethread_Worker initRtMidiEventPort
../src/workers/worklet.singlethread.worker.js:83 WorkletSinglethread_Worker initializing worklet.singlethread.worker
../src/workers/worklet.singlethread.worker.js:118 WorkletSinglethread_Worker waiting on wasm initialization to complete
localhost/:1 Uncaught (in promise) CompileError: WebAssembly.compile(): Compiling function #2560:"init_static_modules" failed: invalid function index: 4294967295 @+1492288
```
it could be that this is still caused by the webpack/react environment - not sure about this.
thanks for any hints.

**Comment by docEdub** [comment]
I get the same error when using vanilla JS. Running `scripts/compile.static.sh` gets past it but then the linker complains about the missing `beosc` opcode.

**Comment by vlazzarini** [comment]
The BeOSC opcode has been removed, maybe that's the reason?

**Comment by docEdub** [comment]
Commenting out the existing csound_wasm.c `init_static_modules` and using the empty implementation instead fixes the normal build for me, but now I'm getting a .orc parse error on the `reshapearray` opcode. Has that been removed, too?
```c
#ifdef INIT_STATIC_MODULES
// extern int init_static_modules(CSOUND *csound);
extern int scansyn_init_(CSOUND *csound);
extern int scansynx_init_(CSOUND *csound);
extern int emugens_init_(CSOUND *csound);
extern int scugens_init_(CSOUND *csound);
extern int beosc_init_(CSOUND *csound);
extern int pvsops_init_(CSOUND *csound);
extern int liveconv_init_(CSOUND *csound);
extern int unsupported_opdoces_init_(CSOUND *csound);
extern int dateops_init_(CSOUND *csound);
#endif

int init_static_modules(CSOUND *csound) { }
// int init_static_modules(CSOUND *csound) {
//   scansyn_init_(csound);
//   scansynx_init_(csound);
//   emugens_init_(csound);
//   scugens_init_(csound);
//   beosc_init_(csound);
//   pvsops_init_(csound);
//   liveconv_init_(csound);
//   dateops_init_(csound);
//   unsupported_opdoces_init_(csound);
//   return 0;
// };
```

**Comment by docb** [comment]
i would guess that this is caused by commenting out emugens_init.

**Comment by vlazzarini** [comment]
I think you only need to comment out what's gone. The `beosc_init_(csound);` seems to be the one.

**Comment by docEdub** [comment]
Ok, ya this works for me now:
```c
#ifdef INIT_STATIC_MODULES
// extern int init_static_modules(CSOUND *csound);
extern int scansyn_init_(CSOUND *csound);
extern int scansynx_init_(CSOUND *csound);
extern int emugens_init_(CSOUND *csound);
extern int scugens_init_(CSOUND *csound);
extern int beosc_init_(CSOUND *csound);
extern int pvsops_init_(CSOUND *csound);
extern int liveconv_init_(CSOUND *csound);
extern int unsupported_opdoces_init_(CSOUND *csound);
extern int dateops_init_(CSOUND *csound);
#endif

// int init_static_modules(CSOUND *csound) { }
int init_static_modules(CSOUND *csound) {
  scansyn_init_(csound);
  scansynx_init_(csound);
  emugens_init_(csound);
  scugens_init_(csound);
  // beosc_init_(csound);
  pvsops_init_(csound);
  liveconv_init_(csound);
  dateops_init_(csound);
  unsupported_opdoces_init_(csound);
  return 0;
};
```

**Comment by vlazzarini** [comment]
Good, does it all build correctly?

**Comment by docEdub** [comment]
Yes, and it looks like commenting out only the `beosc_init_(csound);` line also resolves the error `Compiling function #2560:"init_static_modules" failed: invalid function index ...` in the normal build using `yarn build` as before. No need to run `scripts/compile.static.sh`

**Comment by kunstmusik** [comment]
I pushed a change to remove those lines (looks like I missed those when
merging and updating from 6.16 release code).  Thanks for bring up this
issue.

**Comment by docb** [comment]
thanks this works. but now i have clicks on noteoffs done with turnoff2 (triggered by csoundObj.readScore) which i do not have with CsoundObj.js (6.15) with the identical csd and trigger mechanism. is there something to note here?

**Comment by vlazzarini** [comment]
Can you provide a simple CSD demonstrating that issue? We have made some changes in that area to fix similar issues. I don't know if that code has been pushed to the webaudio branch.

**Comment by vlazzarini** [comment]
I see that the code has been merged alright. I am not sure about this. Can you open a separate ticket with a CSD we can use to reproduce the issue?

**Comment by docb** [comment]
yes ok, i'll try and put into a separate ticket. i think this one can be closed.
